### PR TITLE
Simplify typechecking of Combine and CombineTypes

### PR DESCRIPTION
**Pull request by sjakobi**

### CODE REVIEW

**Comment by EggBaconAndSpam** [comment]
Could you give me a short explanation as to how your change works, and why it is correct? I have to admit that I am slightly lost without some guiding remarks ;)

**Comment by sjakobi** [comment]
> Could you give me a short explanation as to how your change works, and why it is correct? I have to admit that I am slightly lost without some guiding remarks ;)

It's admittedly a bit hacky! `unionWith CombineTypes` is my poor man's [`Data.Semialign.align`](https://hackage.haskell.org/package/semialign-1/docs/Data-Semialign.html#v:align). `CombineTypes` is safe to use as a tag for keys present in both maps, as any `CombineTypes` in either of them have already been eliminated further up via `fmap Dhall.Core.normalize (loop ctx kvs{X,Y})` – we never return a `CombineTypes` from `typeWithA`!

Not sure how to make this less hacky…

Would it be clearer if I used e.g. `Combine` or `ImportAlt` as a tag?

Using [`semialign`](https://hackage.haskell.org/package/semialign) would be clear things up IMHO but would require adding it to our `stack` `extra-deps`. Not sure about the Nix situation…

**Comment by sjakobi** [comment]
I implemented my own `align` now, based on `Data.Map.mergeWithKey`.

For `CombineTypes` I now discovered that it is sufficient to traverse the intersection of the two `Record`s. We never did anything with the other elements but it was a bit hard to see that we didn't need the union. Nice! :)

**Comment by EggBaconAndSpam** [comment]
Neat!

I'll try to have a closer look at this by Monday at the latest (feel free to preempt me, everybody) &ndash; my brain isn't of much use right now.

File: dhall/src/Dhall/Map.hs [hunk 1]
```diff
@@ -381,6 +382,31 @@ unionWith combine (Map mL ksL) (Map mR ksR) = Map m ks
     ks = ksL <|> Prelude.filter (\k -> Data.Set.notMember k setL) ksR
 {-# INLINABLE unionWith #-}
 
+{-| A generalised 'unionWith'.
+
+>>> align Left Left (\k a b -> Right (k, a, b)) (fromList [("A",1),("B",2)]) (singleton "A" 3)
+fromList [("A",Right ("A",1,3)),("B",Left 2)]
+-}
+align
```
[the remark this answers is not in the document]
Comment (sjakobi) [reply dhall/src/Dhall/Map.hs]: Done! :)

**Pull request merged**